### PR TITLE
commit offset to broker after checkpointing (#230)

### PR DESCRIPTION
**Pull request by SuXingLee**
## What is the purpose of the change

Change the guarantee of flink-rocketmq's source from at-most-once to at-least-once when restart flink without using savepoint.

## Brief changelog

Adjust the moment that commit offset to broker.

## Verifying this change

no test

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [x] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/rocketmq/tree/master/test).
- [x] Run `mvn -B clean apache-rat:check findbugs:findbugs checkstyle:checkstyle` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.
- [ ] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

**Comment by SuXingLee** [comment]
@vesense Do yo have time to take a look?

File: rocketmq-flink/src/main/java/org/apache/rocketmq/flink/RocketMQSource.java [hunk 1]
```diff
@@ -84,6 +85,7 @@
     private static final String OFFSETS_STATE_NAME = "topic-partition-offset-states";
 
     private transient volatile boolean restored;
+    private boolean enableCheckpoint;
```
Comment (Jennifer-sarah) [inline rocketmq-flink/src/main/java/org/apache/rocketmq/flink/RocketMQSource.java]: transient variable.

File: rocketmq-flink/src/main/java/org/apache/rocketmq/flink/RocketMQSource.java [hunk 2]
```diff
@@ -291,6 +297,10 @@ public void snapshotState(FunctionSnapshotContext context) throws Exception {
 
         for (Map.Entry<MessageQueue, Long> entry : offsetTable.entrySet()) {
             unionOffsetStates.add(Tuple2.of(entry.getKey(), entry.getValue()));
+            // commit offset to broker after checkpointing
+            if (enableCheckpoint) {
```
Comment (Jennifer-sarah) [inline rocketmq-flink/src/main/java/org/apache/rocketmq/flink/RocketMQSource.java]: logic is wrong here!

**Comment by Jennifer-sarah** [comment]
@SuXingLee, I had left some comments for you. we did a same work for this #229.
btw, are you interested in working with me in hangzhou?

File: rocketmq-flink/src/main/java/org/apache/rocketmq/flink/RocketMQSource.java (hunk 1 above)
Comment (SuXingLee) [reply rocketmq-flink/src/main/java/org/apache/rocketmq/flink/RocketMQSource.java]: no need to set this variable as transient

Comment (Jennifer-sarah) [reply rocketmq-flink/src/main/java/org/apache/rocketmq/flink/RocketMQSource.java]: > no need to set this variable as transient

:+1:

File: rocketmq-flink/src/main/java/org/apache/rocketmq/flink/RocketMQSource.java (hunk 2 above)
Comment (SuXingLee) [reply rocketmq-flink/src/main/java/org/apache/rocketmq/flink/RocketMQSource.java]: you are right, the best way is do this logic when checkpoint compeleted.

**Comment by SuXingLee** [comment]
thank for your review, you code have some problem.see my newest commit.

**Comment by Jennifer-sarah** [comment]
> thank for your review, you code have some problem.see my newest commit.

but we should keep one pr, then fix this problem, right?

**Comment by SuXingLee** [comment]
yes! but i don't know how to directly point out this problem , so use the code , I hope you could understand.

**Comment by Jennifer-sarah** [comment]
> yes! but i don't know how to directly point out this problem , so use the code , I hope you could understand.

???

**Comment by SuXingLee** [comment]
my wx number is 15705102365 ，let's talk directly

**Pull request closed without merging**